### PR TITLE
WPF - Implement UpdateDragCursor

### PR DESCRIPTION
**Pull request by joaompneves**
Implemented the UpdateDragCursor method of the WPF webview.

Tested with: https://jsfiddle.net/tw5f801j/

The drag&drop cursors are retrieved from Windows lib (ole32) and displayed when appropriate.

The implementation still has a minor problem: the cursor flickers, which is more noticeable when the cursor changes to forbidden. I don't know why this happens. If anyone can shed light, that would be much appreciated.

### CODE REVIEW

File: CefSharp.Wpf/ChromiumWebBrowser.cs [hunk 1]
```diff
@@ -787,7 +787,8 @@ void IRenderWebBrowser.UpdateDragCursor(DragOperationsMask operation)
         /// <param name="operation">describes the allowed operation (none, move, copy, link). </param>
         protected virtual void UpdateDragCursor(DragOperationsMask operation)
         {
-            //TODO: Someone should implement this
+            var dragCursor = DragCursorProvider.GetCursor(operation);
+            UiThreadRunSync(() => Mouse.SetCursor(dragCursor));
```
Comment (amaitland) [inline CefSharp.Wpf/ChromiumWebBrowser.cs]: Is there a reason this is run `Sync`? If so please add a comment as to the reason.

File: CefSharp.Wpf/DragCursorProvider.cs [hunk 2]
```diff
@@ -0,0 +1,56 @@
+﻿using System;
```
Comment (amaitland) [inline CefSharp.Wpf/DragCursorProvider.cs]: Needs a license added to the header.

File: CefSharp.Wpf/DragCursorProvider.cs [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+        private static extern IntPtr LoadCursor(IntPtr hInstance, ushort lpCursorName);
+        
+        private static readonly Dictionary<DragOperationsMask, Cursor> DragCursors;
+
+        static DragCursorProvider() 
+        {
+            var library = LoadLibrary("ole32.dll");
+            DragCursors = new Dictionary<DragOperationsMask, Cursor>()
+            {
+                { DragOperationsMask.None, GetCursorFromLib(library, 1) },
```
Comment (amaitland) [inline CefSharp.Wpf/DragCursorProvider.cs]: What's the difference between this `None` cursor and `Cursors.None`? Be better to use the built in one where possible

https://docs.microsoft.com/en-us/dotnet/api/system.windows.input.cursors.none?view=netframework-4.7.2

File: CefSharp.Wpf/DragCursorProvider.cs [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+                { DragOperationsMask.Move, GetCursorFromLib(library, 2) },
```
Comment (amaitland) [inline CefSharp.Wpf/DragCursorProvider.cs]: Is `Move` substantially different from `SizeAll`? Be better to use the built in cursor where possible.

https://docs.microsoft.com/en-us/dotnet/api/system.windows.input.cursors.sizeall?view=netframework-4.7.2#System_Windows_Input_Cursors_SizeAll

**Comment by amaitland** [comment]
> The implementation still has a minor problem: the cursor flickers

How does it behave if you just map the different types to a built in cursor? Is it something to do with the custom cursor?

**Comment by joaompneves** [comment]
> Please provide a list of steps used for testing the different cursor types.

Changed the link with a more self-explanatory example. Just drag the Drag me box over the others.

File: CefSharp.Wpf/ChromiumWebBrowser.cs (hunk 1 above)
Comment (joaompneves) [reply CefSharp.Wpf/ChromiumWebBrowser.cs]: This is an attempt to reduce the flicker, but it still exists.

File: CefSharp.Wpf/DragCursorProvider.cs (hunk 3 above)
Comment (joaompneves) [reply CefSharp.Wpf/DragCursorProvider.cs]: Cursors.None is an invisible cursor as the None in drag&drop is a forbidden one.

File: CefSharp.Wpf/DragCursorProvider.cs (hunk 4 above)
Comment (joaompneves) [reply CefSharp.Wpf/DragCursorProvider.cs]: Drag&drop cursors don't exist in the built-in collection.

These are the ones we get from ole32:
https://i.stack.imgur.com/TffLL.png

**Comment by joaompneves** [comment]
> > The implementation still has a minor problem: the cursor flickers
> 
> How does it behave if you just map the different types to a built in cursor? Is it something to do with the custom cursor?

Behaves the same way.

**Comment by joaompneves** [comment]
Solved the flicker problem. The feature is now OK.

**Comment by joaompneves** [comment]
Turns out the solution is more simple. Just had to store the drag operation result on the update drag cursor and use that in the DragOver event. I have tried this before without success, because the key is to handle the drag over event, otherwise the cursor gets overridden by wpf (default behavior).

**Comment by amaitland** [comment]
Much cleaner thanks 👍 If all goes well testing wise this will be in the `73` release.

**Comment by amaitland** [comment]
Formatting fixes for the html file in https://github.com/cefsharp/CefSharp/commit/ea843f31c172dcd7da4b11d9330b9f9cb2fd8f28

`&` needs to be encoded as `&amp;` for xml doc also

**Pull request merged**